ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/NettyPerfClient.java

```diff
@@ -0,0 +1,501 @@
+    }
+    logger.info("Instantiated NettyPerfClient which will interact with host {}, port {}, uri {} with concurrency {}",
+        this.host, this.port, uri, this.concurrency);
+  }
+
+  /**
+   * Starts the NettyPerfClient.
+   * @throws InterruptedException
+   */
+  protected void start()
+      throws InterruptedException {
+    logger.info("Starting NettyPerfClient");
+    reporter.start();
+    group = new NioEventLoopGroup(concurrency);
+    b.group(group).channel(NioSocketChannel.class).handler(new ChannelInitializer<SocketChannel>() {
+      @Override
+      public void initChannel(SocketChannel ch)
+          throws Exception {
+        ch.pipeline().addLast(new HttpClientCodec()).addLast(new ChunkedWriteHandler()).addLast(new RequestSender());
+      }
+    });
+    logger.info("Connecting to {}:{}", host, port);
+    b.remoteAddress(host, port);
+    perfClientStartTime = System.currentTimeMillis();
+    for (int i = 0; i < concurrency; i++) {
+      b.connect().addListener(channelConnectListener);
+    }
+    logger.info("Created {} channel(s)", concurrency);
+    logger.info("NettyPerfClient started");
+  }
+
+  /**
+   * Shuts down the NettyPerfClient.
+   */
+  protected void shutdown() {
+    logger.info("Shutting down NettyPerfClient");
+    isRunning = false;
+    group.shutdownGracefully();
+    long totalRunTimeInMs = System.currentTimeMillis() - perfClientStartTime;
+    try {
+      if (!group.awaitTermination(5, TimeUnit.SECONDS)) {
+        logger.error("Netty worker did not shutdown within timeout");
+      } else {
+        logger.info("NettyPerfClient shutdown complete");
+      }
+    } catch (InterruptedException e) {
+      logger.error("NettyPerfClient shutdown interrupted", e);
+    } finally {
+      logger.info("Executed for approximately {} s and sent {} requests ({} requests/sec)",
+          (float) totalRunTimeInMs / (float) Time.MsPerSec, totalRequestCount.get(),
+          (float) totalRequestCount.get() * (float) Time.MsPerSec / (float) totalRunTimeInMs);
+      Snapshot rttStatsSnapshot = perfClientMetrics.requestRoundTripTimeInMs.getSnapshot();
+      logger.info("RTT stats: Min - {} ms, Mean - {} ms, Max - {} ms", rttStatsSnapshot.getMin(),
+          rttStatsSnapshot.getMean(), rttStatsSnapshot.getMax());
+      logger.info("RTT stats: 95th percentile - {} ms, 99th percentile - {} ms, 999th percentile - {} ms",
+          rttStatsSnapshot.get95thPercentile(), rttStatsSnapshot.get99thPercentile(),
+          rttStatsSnapshot.get999thPercentile());
+      reporter.stop();
+      shutdownLatch.countDown();
+    }
+  }
+
+  /**
+   * Blocking function to wait on the NettyPerfClient shutting down.
+   * @throws InterruptedException
+   */
+  protected void awaitShutdown()
+      throws InterruptedException {
+    shutdownLatch.await();
+  }
+
+  /**
+   * Custom handler that sends out the request and receives responses.
+   */
+  private class RequestSender extends SimpleChannelInboundHandler<Object> {
```

Comment:
It is an inbound handler because it has to process responses. I'll change the name to avoid the confusion.

Any handler may do channel writes - it does not need to be an outbound handler for that.